client/lib/checkout/index.js

```diff
@@ -10,8 +14,61 @@ import {
 	hasProduct,
 } from 'lib/cart-values/cart-items';
 import { managePurchase } from 'me/purchases/paths';
+import {
+	UPGRADE_INTENT_PLUGINS,
+	UPGRADE_INTENT_INSTALL_PLUGIN,
+	UPGRADE_INTENT_THEMES,
+	UPGRADE_INTENT_INSTALL_THEME,
+} from 'lib/checkout/constants';
+import { decodeURIComponentIfValid, isExternal } from 'lib/url';
+
+const isValidValue = url => typeof url === 'string' && url;
+
+export function isValidUpgradeIntent( upgradeIntent ) {
+	switch ( upgradeIntent ) {
+		case UPGRADE_INTENT_PLUGINS:
+		case UPGRADE_INTENT_INSTALL_PLUGIN:
+		case UPGRADE_INTENT_THEMES:
+		case UPGRADE_INTENT_INSTALL_THEME:
+			return true;
+	}
+	return false;
+}
+
+export function parseRedirectToChain( rawUrl, includeOriginalUrl = true ) {
+	const redirectChain = [];
+	const url = decodeURIComponentIfValid( rawUrl );
 
-export function getExitCheckoutUrl( cart, siteSlug ) {
+	if ( includeOriginalUrl && isValidValue( url ) ) {
+		redirectChain.push( url );
+	}
+
+	const parseUrlAndPushToChain = currentUrl => {
+		const {
+			query: { redirect_to },
+		} = parse( currentUrl, true );
+		if ( isValidValue( redirect_to ) ) {
+			redirectChain.push( redirect_to );
+			parseUrlAndPushToChain( decodeURIComponentIfValid( redirect_to ) );
+		}
+	};
+
+	parseUrlAndPushToChain( url );
+	return redirectChain;
+}
+
+export function getValidDeepRedirectTo( redirectTo ) {
+	const redirectChain = parseRedirectToChain( redirectTo );
+	if (
+		Array.isArray( redirectChain ) &&
+		redirectChain.length &&
+		! isExternal( redirectChain[ redirectChain.length - 1 ] )
```

Comment:
Hm I've seen this responsibility in consumers of `redirect_to`, but this is actually pretty smart since it's a pretty standard check. If we need raw value, we can always introduce `getDangerousRawDeepRedirectTo` at some point in the future.